compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/IntegerDivRemNode.java

```diff
@@ -83,6 +85,92 @@ public final Type getType() {
+        int shiftNum = nums.getRight().intValue();
+        ValueNode value;
+        if (bitSize == Integer.SIZE) {
+            value = new MulNode(new SignExtendNode(forX, Long.SIZE), ConstantNode.forLong(magicNum));
+            if (magicNum < 0) {
+                value = new RightShiftNode(value, ConstantNode.forInt(bitSize));
+                value = new AddNode(forX, new NarrowNode(value, bitSize));
+                if (shiftNum != 0) {
+                    value = new RightShiftNode(value, ConstantNode.forInt(shiftNum));
+                }
+            } else {
+                value = new RightShiftNode(value, ConstantNode.forInt(bitSize + shiftNum));
+                value = new NarrowNode(value, bitSize);
+            }
+        } else {
+            assert bitSize == Long.SIZE;
+            value = new IntegerMulHighNode(forX, ConstantNode.forLong(magicNum));
+            if (magicNum < 0) {
+                value = new AddNode(forX, value);
+            }
+            if (shiftNum != 0) {
+                value = new RightShiftNode(value, ConstantNode.forInt(shiftNum));
+            }
+        }
+
+        ValueNode subValue = new RightShiftNode(forX, ConstantNode.forInt(bitSize - 1));
+        if (isDivisorPositive) {
+            return new SubNode(value, subValue);
+        }
+        return new SubNode(subValue, value);
+    }
+
+    /**
+     * Borrowed from Hacker's Delight by Henry S. Warren, Jr.
+     */
+    private static Pair<Long, Integer> magicDivideConstants(long divisor, int size) {
+        if (divisor == 0 || divisor == 1) {
+            return null;
+        }
+
+        final long twoW = 1L << (size - 1);                // 2 ^ (size - 1).
+        long t = twoW + (divisor >>> (size - 1));
+        long ad = Math.abs(divisor);
+        long anc = t - 1 - Long.remainderUnsigned(t, ad);  // Absolute value of nc.
+        long q1 = Long.divideUnsigned(twoW, anc);          // Init. q1 = 2**p/|nc|.
+        long r1 = Long.remainderUnsigned(twoW, anc);       // Init. r1 = rem(2**p, |nc|).
+        long q2 = Long.divideUnsigned(twoW, ad);           // Init. q2 = 2**p/|d|.
+        long r2 = Long.remainderUnsigned(twoW, ad);        // Init. r2 = rem(2**p, |d|).
+        long delta;
+
+        int p = size - 1;                                  // Init. p.
+        do {
+            p = p + 1;
+            q1 = 2 * q1;                                   // Update q1 = 2**p/|nc|.
+            r1 = 2 * r1;                                   // Update r1 = rem(2**p, |nc|).
+            if (Long.compareUnsigned(r1, anc) >= 0) {      // Must be an unsigned comparison.
+                q1 = q1 + 1;
+                r1 = r1 - anc;
+            }
+            q2 = 2 * q2;                                   // Update q2 = 2**p/|d|.
+            r2 = 2 * r2;                                   // Update r2 = rem(2**p, |d|).
+            if (r2 >= ad) {                                // Must be an unsigned comparison.
```

Comment:
This should be unsigned comparison like line 153, right?

Reply:
I update here and use the unsigned comparison instead. Thanks!